src/vw_slim_model/src/output_task.cpp

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <inference/direct_input_pipe.hpp>
+#include <inference/feature_error.hpp>
+#include <vw_common/error.hpp>
+#include <vw_slim_model/output_task.hpp>
+
+#pragma warning(push)
+#pragma warning(disable : 26451)  // Arithmetic overflow warnings in hash.h.
+#pragma warning(disable : 26495)  // Uninitialized members and variables seemingly everywhere.
+#pragma warning(disable : 26812)  // Enums used throughout instad of enum class.
+#include "array_parameters.h"
+#include "example_predict_builder.h"
+#include "vw_slim_predict.h"
+#include "vw_slim_return_codes.h"
+#pragma warning(pop)
+
+typedef vw_slim::vw_predict<::sparse_parameters> VWModel;
+typedef ::safe_example_predict VWExample;
+
+namespace {
+tl::expected<inference::TypeDescriptor, std::error_code> ActionTypeToType(
+    const resonance_vw::ActionType atype) noexcept {
+    switch (atype) {
+        case resonance_vw::ActionType::Float:
+            return inference::TypeDescriptor::Create<inference::Tensor<float>>();
+        case resonance_vw::ActionType::Int:
+            return inference::TypeDescriptor::Create<inference::Tensor<int>>();
+        case resonance_vw::ActionType::String:
+            return inference::TypeDescriptor::Create<inference::Tensor<std::string>>();
+    }
+    return tl::make_unexpected(inference::make_feature_error(inference::feature_errc::type_unsupported));
+}
+
+size_t CountOfActions(std::shared_ptr<resonance_vw::Actions> const& actions) noexcept {
+    switch (actions->Type()) {
+        case resonance_vw::ActionType::Int:
+            return actions->GetIntActions().value().size();
+        case resonance_vw::ActionType::String:
+            return actions->GetStringActions().value().size();
+    }
+    return 0;
+}
+}  // namespace
+
+namespace resonance_vw {
+
+class OutputTask::IPokerImpl : public IPoker {
+   protected:
+    const std::shared_ptr<VWModel> m_model;
+    const std::shared_ptr<VWExample> m_example;
+
+   public:
+    IPokerImpl(std::shared_ptr<void> vwModel, std::shared_ptr<void> vwExample);
+    ~IPokerImpl() = default;
+};
+
+OutputTask::IPokerImpl::IPokerImpl(std::shared_ptr<void> vwModel, std::shared_ptr<void> vwExample)
+    : m_model(std::static_pointer_cast<VWModel>(vwModel)), m_example(std::static_pointer_cast<VWExample>(vwExample)) {}
+
+OutputTask::IPoker::IPoker() {}
```

Comment:
The structure seems a little odd and maybe its naming for me, but there is an OutputTask, which makes me think of the C# Task class or at the very least something that I could run. But in reality it implements the IPoker interface, so in the end it acts as an poker and doesnt look like a task. It then implements an internal Poker class that is specific to regression or whatever the algorithm is -- why not just have a RegressionPoker class, RecommendationPoker class, etc?

Reply:
> But in reality it implements the IPoker interface

Well, the tasks do not implement the interface, they can return objects that implement this `IPoker` interface. Note also that the entire `IPoker` interface is entirely private. It is not a concept publicly usable by the user. The task is the only thing the user sees. Architecturally, pokers are stateful objects (they own the `m_example` that is built out of the inputs of a particular bound output) whereas the tasks are stateless.

It is analogous, somewhat, to the relationship between `inference::Model` and `inference::ValueUpdater`.

The naming though might need some work. I agree that `Task` is not a great name, though I am not excited about `Poker` being a name. I can explain `OutputTask` fairly easily: it is the task for which you are employing the output of the model. I can't explain `Poker` without getting into a lot of details that the user, in principle, should never have to care about.

So alternatives to task. I guess the idea is, like we have a schema builder for the inputs, we have something that shows out the output is to be interpreted. `OutputInterpreter`? `OutputDescriptor`? Something else? What you think @singlis?

Reply:
Thanks for the explanation @TomFinley. I think initially for me it was weird, but after I saw more how it was used, it makes sense. So please disregard.